pkg/personname/info.go

```diff
@@ -0,0 +1,202 @@
+func (group pnGroup) String() string {
+	switch group {
+	case 0:
+		return "Alphabetic"
+	case 1:
+		return "Ideographic"
+	case 2:
+		return "Phonetic"
+	default:
+		panic(fmt.Errorf("bad pnGroup value: %v", int(group)))
+	}
+}
+
+// Enum values for pnGroup
+const (
+	pnGroupAlphabetic pnGroup = iota
+	pnGroupIdeographic
+	pnGroupPhonetic
+)
+
+// Info holds information from an element with a "PN" VR. See the "PN"
+// entry at: http://dicom.nema.org/dicom/2013/output/chtml/part05/sect_6.2.html
+//
+// PN values are split into three groups which represent three different ways to
+// represent a name:
+//
+// - Alphabetic: How a name is formally spelled using a Phonetic alphabet.
+// - Ideographic: How a name is represented using ideograms / ideographs.
+// - Phonetic: How a name is pronounced.
+//
+// Each of these groups can be inspected to access their individual segments (family
+// name, Given name, etc.)
+type Info struct {
+	// Alphabetic group information about the Alphabetic group.
+	Alphabetic GroupInfo
+	// Ideographic group information about the Ideographic group.
+	Ideographic GroupInfo
+	// Phonetic group information about the Phonetic group.
+	Phonetic GroupInfo
+
+	// NoNullSeparators will remove repeated separators around null groups when
+	// calling DCM() if set to true.
+	NoNullSeparators bool
+}
+
+// WithNullSeparators returns a new Info object that will keep trailing separators
+// that surround both null groups AND group segments: (ex: 'Potter^Harry^^^==').
+func (info Info) WithNullSeparators() Info {
+	info.NoNullSeparators = false
+	info.Alphabetic.NoNullSeparators = false
+	info.Ideographic.NoNullSeparators = false
+	info.Phonetic.NoNullSeparators = false
+
+	return info
+}
+
+// WithoutNullSeparators returns a new Info object that will remove trailing
+// separators that surround both null groups AND group segments:
+// (ex: 'Potter^Harry').
+func (info Info) WithoutNullSeparators() Info {
+	info.NoNullSeparators = true
+	info.Alphabetic.NoNullSeparators = true
+	info.Ideographic.NoNullSeparators = true
+	info.Phonetic.NoNullSeparators = true
+
+	return info
+}
+
+// dcmRemoveNullStrings removes null group strings from the list of groups to be
+// rendered.
+func (info Info) dcmRemoveNullStrings(groupStrings []string) []string {
+	// If we are not removing null separators, return immediately.
+	if !info.NoNullSeparators {
+		return groupStrings
+	}
+
+	// If we are removing trailing emtpy, we are going to check if the last group is
+	// empty, and if it is, trim it off the end until we hit a non-empty slice.
+	//
+	// If we were to move front-to-back instead of back-to-front, we would get a bad
+	// result if the middle group is blank, but the third group is not, as we would
+	// bail on the middle group, and not end up rendering the last.
+	//
+	// 'Harry^Potter==Phonetic^Spelling is a valid PN value, which we need to account
+	// for.
+
+	// Start with the sliceOut point containing all three groups
+	sliceOut := 3
```

Comment:
should we just use len(groupStrings) here? I guess it'll always be 3.

Reply:
That might be a little safer -- Since it should always be 3 I was thinking we could save the length check, but I tend to think it's better to err on the side of safety as well. Will change.